File: coding/extras/design_patterns/observer.cpp
#include<iostream>
#include<bits/stdc++.h>

using namespace std;

class Observer{
    public:
    virtual ~Observer() = default;
    virtual void update() = 0;
};

class Subject{
    public:
        void attach(const shared_ptr<Observer>& ob){
            observers.push_back(ob);
        }

        void detach(const shared_ptr<Observer>& ob){
            observers.erase(remove(observers.begin() , observers.end(), ob),observers.end());
        }

        void notify(){
            for(auto o : observers){
                o->update();
            }
        }

    private:
        vector<shared_ptr<Observer>> observers;
};

class Stock: public Subject{
    public:
        void setPrice(float p){
            price = p;
            notify();
        }

        float getPrice(){
            return price;
        }
    private:
        float price = 0.0f;
};

class Investor : public Observer{
    public:
        Investor(string n , shared_ptr<Stock> s) : name(n) , stock(s){}

        void update(){
            cout<<" Investor : "<<name<<" notified for stock price  "<<stock->getPrice()<<endl;
        }
    private:
        string name;
        shared_ptr<Stock> stock;
};

int main(){
    auto st = make_shared<Stock>();

    auto in1 = make_shared<Investor>("Alice",st);
    auto in2 = make_shared<Investor>("Bob ",st);

    st->attach(in1);
    st->attach(in2);

    st->setPrice(101.12);
    st->setPrice(104.31);

    st->detach(in1);
    st->setPrice(150.9);
}

File: golang/day1/crud/observer.go
package main

import (
	"fmt"
	"sync"
)

// Observer interface
type Observer interface {
	Update(price float64)
}

// Subject struct
type Subject struct {
	observers map[Observer]struct{}
	mu        sync.Mutex
}

func (s *Subject) Attach(o Observer) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.observers[o] = struct{}{}
}

func (s *Subject) Detach(o Observer) {
	s.mu.Lock()
	defer s.mu.Unlock()

	delete(s.observers, o)
}

func (s *Subject) Notify(price float64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for o := range s.observers {
		o.Update(price)
	}
}

// Stock struct
type Stock struct {
	Subject
	price float64
}

func NewStock() *Stock {
	return &Stock{
		Subject: Subject{observers: make(map[Observer]struct{})},
	}
}

func (s *Stock) SetPrice(price float64) {
	s.price = price
	s.Notify(price)
}

// Investor struct
type Investor struct {
	name  string
	stock *Stock
}

func (i *Investor) Update(price float64) {
	fmt.Printf("Investor %s notified of stock price: %.2f\n", i.name, i.stock.price)
}

func main() {
	stock := NewStock()

	investor1 := &Investor{name: "Alice", stock: stock}
	investor2 := &Investor{name: "Bob", stock: stock}

	stock.Attach(investor1)
	stock.Attach(investor2)

	stock.SetPrice(101.12)
	stock.SetPrice(104.31)

	stock.Detach(investor1)
	stock.SetPrice(150.9)
}
